designs/v1beta1-api.md

```diff
@@ -0,0 +1,341 @@
+    1. Neutral Node configuration-based fields that affect the **compatibility between Nodes and Pods during scheduling** (e.g. requirements, taints, labels)
+    2. Neutral behavior-based fields for configuring Karpenter’s scheduling and deprovisioning decision-making
+2. `NodeClass`
+    1. Cloudprovider-specific Node configuration-based fields that affect launch and bootstrap process for that Node including: configuring startup scripts, volume mappings, metadata settings, etc.
+    2. Cloudprovider-specific behavior-based fields for configuring Karpenter’s scheduling and deprovisioning decision-making (e.g. interruption-based disruption, allocation strategy)
+3. `NodeClaim`
+    1. A Karpenter management object that fully manages the lifecycle of a single node including: configuring and launching the node, monitoring the node health (including disruption conditions), and handling the deprovisioning and termination of the node
+
+### `karpenter.sh/NodePool`
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  template:
+    metadata:
+      labels:
+        billing-team: my-team
+      annotations:
+        example.com/owner: "my-team"
+    spec:
+      nodeClass:
+        name: default
+        kind: NodeClass
+        apiVersion: compute.k8s.aws/v1beta1
+      taints:
+        - key: example.com/special-taint
+          effect: NoSchedule
+      startupTaints:
+        - key: example.com/another-taint
+          effect: NoSchedule
+      requirements:
+        - key: "karpenter.k8s.aws/instance-category"
+          operator: In
+          values: ["c", "m", "r"]
+      resources:
+        requests:
+          cpu: "1"
+          memory: "100Mi"
+      kubeletConfiguration:
+        clusterDNS: ["10.0.1.100"]
+        containerRuntime: containerd
+        systemReserved:
+          cpu: 100m
+          memory: 100Mi
+          ephemeral-storage: 1Gi
+        kubeReserved:
+          cpu: 200m
+          memory: 100Mi
+          ephemeral-storage: 3Gi
+        evictionHard:
+          memory.available: 5%
+          nodefs.available: 10%
+          nodefs.inodesFree: 10%
+        evictionSoft:
+          memory.available: 500Mi
+          nodefs.available: 15%
+          nodefs.inodesFree: 15%
+        evictionSoftGracePeriod:
+          memory.available: 1m
+          nodefs.available: 1m30s
+          nodefs.inodesFree: 2m
+        evictionMaxPodGracePeriod: 60
+        imageGCHighThresholdPercent: 85
+        imageGCLowThresholdPercent: 80
+        cpuCFSQuota: true
+        podsPerCore: 2
+        maxPods: 20
+  deprovisioning:
+    consolidationTTL: 10m
+    consolidationPolicy: WhenEmpty | WhenUnderutilized
+    expirationTTL: 30d
+    uninitializedTTL: 20m
+  weight: 10
+  limits:
+    cpu: "1000"
+    memory: 1000Gi
+status:
+  resources:
+     cpu: "2"
+     memory: "100Mi"
+     ephemeral-storage: "100Gi"
+```
+
+### `compute.k8s.aws/NodeClass`
+
+```
+apiVersion: compute.k8s.aws/v1beta1
```

Comment:
> > I think something nice would be "version" similar to LaunchTemplate version. This allows Cluster Operators to rollout and rollback changes to nodes safely and have the necessary controls when managing many NodeClasses. Ideally each NodeClasses version should be immutable. Right now it is difficult to expand on `NodeClass` for nodes outside of AWS. It may be valuable to split NodeClasses, I don't have a good name, such that we have something that is at the cloud provider level and one that just defines nodes specification specific to a cloud provider. (Similar to gatewayclass and gateway in Gateway API). The Cloud provider layer resource defines the controller for creating nodes for that cloud provider and the nodeclass themselves work in the context of each cloud provider.
> 
> I'm still thinking on the other comment that you made around the Gateway API, but, for this one, the idea was that the `NodeClass` _is_ the cloudprovider extensibility mechanism; it's effectively the interface to define how Karpenter can actually bring up a Node in a given Cloudprovider. If another Cloudprovider came along and wanted to support Karpenter, I would expect that they would create their own NodeClass and this NodeClass would define its own interface with how to launch Nodes in that Cloudprovider.
> 
> Generally, we wanted to avoid creating too many disparate concepts that have to be individually reasoned about (i.e. more API kinds). We effectively have taken the stance that if something is common enough that it should be useful to _all_ cloudproviders, then it should be hoisted up into the `NodePool`; otherwise, it belongs within the cloudprovider-specific `NodeClass`.
> 
> Is there a specific use-case that you have in mind where you think splitting out the details of the `NodeClass` could be useful beyond hoisting common cross-cloudprovider values into the `NodePool`?

Sorry moving this to here instead because github threading is kinda bad on main comment. 

Ah Ok I think i am getting the picture more now thanks for the explanation. I think I agree with you NodeClass is a good interface and I think likely don't need to split Nodeclasses out. 

I wonder if there is value as part of this to separate out Karpenter's AWS NodeClass controller out from the main project and reorganize Karpenter vs Karpenter core. Karpenter-core feels more like a library atm. I think there was part of my confusion in understanding how NodeClass may work since right now AWS' pieces feels so integrated with Karpenter as a whole.

Reply:
> I wonder if there is value as part of this to separate out Karpenter's AWS NodeClass controller out from the main project and reorganize Karpenter vs Karpenter core

It's funny because we've thought about this at a few different points in time and I think in a lot of ways we are really close to being able to do this right now. I think it really depends on the value that we get from separating it out like this. When there are more cloudproviders in place, I could see karpenter-core being a "common controller" that handles machine scheduling and hands off Machine launch and deprovisioning process off to the Cloudprovider through Create/Delete on the API server.

That pretty much leaves InstanceTypes as the final piece to separate out into a separate controller, which we have also thought about at different points in time could be communicated through a `compute.k8s.aws/InstanceType` CRD. 

All of this is of course ignoring the fact that you could just write the CloudProvider interface to be gRPC instead of library/in-code without even re-architecting anything